File: Task1/src/py/keygen.py
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import rsa_core

# yrq 0627 存到Task1/
TASK1_DIR = Path(__file__).resolve().parents[2]


def main(argv: list[str]) -> int:
    key_size = 1024
    if len(argv) >= 2:
        key_size = int(argv[1])
        if key_size < 16 or key_size % 2 != 0:
            print("key_size_bits must be an even number >= 16", file=sys.stderr)
            return 1

    keys = rsa_core.generate_rsa_keypair(key_size)
    rsa_core.save_public_key(keys["n"], keys["e"], TASK1_DIR / "RSA_Public_Key.txt")
    rsa_core.save_private_key(keys["n"], keys["d"], TASK1_DIR / "RSA_Secret_Key.txt")
    rsa_core.save_decimal(keys["n"], TASK1_DIR / "RSA_Moduler.txt")
    rsa_core.save_decimal(keys["p"], TASK1_DIR / "RSA_p.txt")
    rsa_core.save_decimal(keys["q"], TASK1_DIR / "RSA_q.txt")
    print(f"RSA key generated: {keys['n'].bit_length()} bits")
    print("Public key, private key, modulus, and prime numbers have been saved to Task1/")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: Task1/src/cpp/keygen.cpp
#include "rsa_utils.h"
#include <gmpxx.h>
#include <iostream>
#define DEFAULT_KEY_SIZE 1024
namespace {
auto print_usage(const char *program) -> void {
	std::cerr << "Usage: " << program
			  << " [key_size_bits] [public_key_file] [private_key_file]\n";
}

auto parse_key_size(const std::string &value) -> std::size_t {
	if (value.empty() || value[0] == '-')
		throw std::runtime_error("key_size_bits must be a positive integer");

	std::size_t parsed_chars = 0;
	const auto key_size = std::stoul(value, &parsed_chars);
	if (parsed_chars != value.size())
		throw std::runtime_error("key_size_bits must be a positive integer");
	if (key_size < 16 || key_size % 2 != 0)
		throw std::runtime_error("key_size_bits must be an even number >= 16");
	return key_size;
}
} // namespace

auto main(int argc, char *argv[]) -> int {
    if (argc > 4) {
        print_usage(argv[0]);
        return 1;
    }

    std::size_t key_size = DEFAULT_KEY_SIZE;
    if (argc >= 2)
        key_size = parse_key_size(argv[1]);

    const std::string public_key_file =
        argc >= 3 ? argv[2]
                    : textbookRSA::resolve_key_path("../../RSA_Public_Key.txt");
    const std::string private_key_file =
        argc >= 4 ? argv[3]
                    : textbookRSA::resolve_key_path("../../RSA_Secret_Key.txt");

    mpz_class n, e, d, p, q;
    textbookRSA::generate_keys(key_size, n, e, d, p, q);
    textbookRSA::save_public_key(n, e, public_key_file);
    textbookRSA::save_private_key(n, d, private_key_file);
    textbookRSA::save_decimal(
        n, textbookRSA::resolve_key_path("../../RSA_Moduler.txt"));
    textbookRSA::save_decimal(
        p, textbookRSA::resolve_key_path("../../RSA_p.txt"));
    textbookRSA::save_decimal(
        q, textbookRSA::resolve_key_path("../../RSA_q.txt"));

    std::cout << "Generated Public Key, Private Key, Big Prime and Modulus" << std::endl;
    return 0;

}
